[FSDP] Do not check fwd order in eval mode

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#77195 [FSDP] Do not check fwd order in eval mode**

Fixes https://github.com/pytorch/pytorch/issues/76949.

**Overview**
If we check the forward execution order in eval mode, where there is no backward pass (namely, no post-backward), the execution order data will not be reset to mark the end of the iteration. This results in false positive warnings of deviating execution order.

This PR removes any checking during eval mode.
- I believe this to be the most reasonable solution since without a backward pass, there is no reliable way to know the end of an iteration.
- In training, the reset point is in `_wait_for_post_backward()`.

**Additional Minor Changes**
In addition, this PR
- removes the unnecessary `reset()` call in `_post_backward_hook()` when in `no_sync()` since `reset()` will still be called in `_wait_for_post_backward()`;
- moves the order check to be called even if the forward `_rebuild_full_params()` will not call all-gather (i.e. if the parameter is not sharded or if it is already rebuilt) since more broadly, the forward order is being validated, not just the all-gather order; and
    - Note: This is useful if this iteration is in train mode and the last iteration was in eval mode. In that case, the root does not free its full parameters and will not all-gather them in this iteration, which would give a false positive warning of the order deviating if we only tracked actual all-gathers.
- improves the warning/error messaging.

Differential Revision: [D36293454](https://our.internmc.facebook.com/intern/diff/D36293454)

Commit:
[FSDP] Do not check fwd order in eval mode

[ghstack-poisoned]

Commit:
Update on "[FSDP] Do not check fwd order in eval mode"


Fixes https://github.com/pytorch/pytorch/issues/76949.

If we check the forward execution order in eval mode, where there is no backward pass (namely, no post-backward), the execution order data will not be reset to mark the end of the iteration. This results in false positive warnings of deviating execution order.

This PR removes any checking during eval mode.
- I believe this to be the most reasonable solution since without a backward pass, there is no reliable way to know the end of an iteration.
- In training, the two reset points are in `_wait_for_post_backward()` when not in `no_sync()` and in the root's `_post_backward()` when in `no_sync()`, both of which should occur later in time than the last `forward()`.

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36293454).

Commit:
Update on "[FSDP] Do not check fwd order in eval mode"


Fixes https://github.com/pytorch/pytorch/issues/76949.

If we check the forward execution order in eval mode, where there is no backward pass (namely, no post-backward), the execution order data will not be reset to mark the end of the iteration. This results in false positive warnings of deviating execution order.

This PR removes any checking during eval mode.
- I believe this to be the most reasonable solution since without a backward pass, there is no reliable way to know the end of an iteration.
- In training, the two reset points are in `_wait_for_post_backward()` when not in `no_sync()` and in the root's `_post_backward()` when in `no_sync()`, both of which should occur later in time than the last `forward()`.

Differential Revision: [D36293454](https://our.internmc.facebook.com/intern/diff/D36293454)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36293454).

rohan-varma reviewed (commented):
I think this makes sense, but is an alternative solution to reset the end of iteration information when we get to the _next_ forward? It seems like we can just check if we haven't reset the previous forward's end of iteration, and then reset it as needed? Just a suggestion in case we haven't considered.

awgu commented:
> I think this makes sense, but is an alternative solution to reset the end of iteration information when we get to the _next_ forward? It seems like we can just check if we haven't reset the previous forward's end of iteration, and then reset it as needed? Just a suggestion in case we haven't considered.

I think we want FSDP to support multiple forward passes with the same module before running the backward pass (since this was part of the constraints brought up motivating the current `FlatParameter` and `FlattenParamsWrapper` design). In that case, the backward pass is the logical signal that an iteration ends. There is no way to tell if a given `forward()` will be the last `forward()` of the iteration.

zhaojuanmao commented:
regarding "In training, the two reset points are in _wait_for_post_backward() when not in no_sync() and in the root's _post_backward() when in no_sync(), both of which should occur later in time than the last forward()."

I missed this in previous PRs, even when in no_sync() mode, _wait_for_post_backward() is still queued and called, because that is the final call back for the whole backward pass no matter no_sync() mode or not. 

in this case, we can reset the execution order data in _wait_for_post_backward for both with/without no_sync()?

awgu commented:
> in this case, we can reset the execution order data in _wait_for_post_backward for both with/without no_sync()?

Great point! I have fixed this.

Commit:
Update on "[FSDP] Do not check fwd order in eval mode"


Fixes https://github.com/pytorch/pytorch/issues/76949.

If we check the forward execution order in eval mode, where there is no backward pass (namely, no post-backward), the execution order data will not be reset to mark the end of the iteration. This results in false positive warnings of deviating execution order.

This PR removes any checking during eval mode.
- I believe this to be the most reasonable solution since without a backward pass, there is no reliable way to know the end of an iteration.
- In training, the two reset points are in `_wait_for_post_backward()` when not in `no_sync()` and in the root's `_post_backward()` when in `no_sync()`, both of which should occur later in time than the last `forward()`.

Differential Revision: [D36293454](https://our.internmc.facebook.com/intern/diff/D36293454)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36293454).

zhaojuanmao reviewed (commented):
looks good!

Commit:
Update on "[FSDP] Do not check fwd order in eval mode"


Fixes https://github.com/pytorch/pytorch/issues/76949.

**Overview**
If we check the forward execution order in eval mode, where there is no backward pass (namely, no post-backward), the execution order data will not be reset to mark the end of the iteration. This results in false positive warnings of deviating execution order.

This PR removes any checking during eval mode.
- I believe this to be the most reasonable solution since without a backward pass, there is no reliable way to know the end of an iteration.
- In training, the reset point is in `_wait_for_post_backward()`.

**Additional Minor Changes**
In addition, this PR
- removes the unnecessary `reset()` call in `_post_backward_hook()` when in `no_sync()` since `reset()` will still be called in `_wait_for_post_backward()`;
- moves the order check to be called even if the forward `_rebuild_full_params()` will not call all-gather (i.e. if the parameter is not sharded or if it is already rebuilt) since more broadly, the forward order is being validated, not just the all-gather order; and
    - Note: This is useful if this iteration is in train mode and the last iteration was in eval mode. In that case, the root does not free its full parameters and will not all-gather them in this iteration, which would give a false positive warning of the order deviating if we only tracked actual all-gathers.
- improves the warning/error messaging.

Differential Revision: [D36293454](https://our.internmc.facebook.com/intern/diff/D36293454)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36293454).

Commit:
Update on "[FSDP] Do not check fwd order in eval mode"


Fixes https://github.com/pytorch/pytorch/issues/76949.

**Overview**
If we check the forward execution order in eval mode, where there is no backward pass (namely, no post-backward), the execution order data will not be reset to mark the end of the iteration. This results in false positive warnings of deviating execution order.

This PR removes any checking during eval mode.
- I believe this to be the most reasonable solution since without a backward pass, there is no reliable way to know the end of an iteration.
- In training, the reset point is in `_wait_for_post_backward()`.

**Additional Minor Changes**
In addition, this PR
- removes the unnecessary `reset()` call in `_post_backward_hook()` when in `no_sync()` since `reset()` will still be called in `_wait_for_post_backward()`;
- moves the order check to be called even if the forward `_rebuild_full_params()` will not call all-gather (i.e. if the parameter is not sharded or if it is already rebuilt) since more broadly, the forward order is being validated, not just the all-gather order; and
    - Note: This is useful if this iteration is in train mode and the last iteration was in eval mode. In that case, the root does not free its full parameters and will not all-gather them in this iteration, which would give a false positive warning of the order deviating if we only tracked actual all-gathers.
- improves the warning/error messaging.

Differential Revision: [D36293454](https://our.internmc.facebook.com/intern/diff/D36293454)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36293454).

awgu commented:
Build failure for `pytorch-xla-linux-bionic-py3.7-clang8` should be unrelated:
```
2022-05-11T17:34:00.8032046Z In file included from /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_bridge.cpp:11:
2022-05-11T17:34:00.8032525Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/tensor_impl.h:33:40: error: only virtual member functions can be marked 'override'
2022-05-11T17:34:00.8032871Z   at::IntArrayRef sizes_custom() const override;
2022-05-11T17:34:00.8033077Z                                        ^~~~~~~~
2022-05-11T17:34:00.8033472Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/tensor_impl.h:34:42: error: only virtual member functions can be marked 'override'
2022-05-11T17:34:00.8033795Z   at::IntArrayRef strides_custom() const override;
2022-05-11T17:34:00.8034006Z                                          ^~~~~~~~
2022-05-11T17:34:00.8034404Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/tensor_impl.h:36:30: error: only virtual member functions can be marked 'override'
2022-05-11T17:34:00.8034764Z   int64_t dim_custom() const override;
2022-05-11T17:34:00.8034951Z                              ^~~~~~~~
2022-05-11T17:34:00.8035344Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/tensor_impl.h:38:32: error: only virtual member functions can be marked 'override'
2022-05-11T17:34:00.8035646Z   int64_t numel_custom() const override;
2022-05-11T17:34:00.8035850Z                                ^~~~~~~~
2022-05-11T17:34:00.8036022Z 4 errors generated.
```

awgu commented:
@pytorchbot merge this please
